src/loaders/masi16_load.c

```diff
@@ -46,7 +46,154 @@ static int masi16_test(HIO_HANDLE *f, char *t, const int start)
 }
 
 
-/* FIXME: effects translation */
+static void masi16_translate_effect(struct xmp_event *event, uint8 effect, uint8 param)
+{
+	switch (effect) {
+	case 1:					/* Fine Volume Slide Up */
+		event->fxt = FX_F_VSLIDE_UP;
+		event->fxp = param;
+		break;
+	case 2:					/* Volume Slide Up */
+		event->fxt = FX_VOLSLIDE_UP;
+		event->fxp = param;
+		break;
+	case 3:					/* Fine Volume Slide Down */
+		event->fxt = FX_F_VSLIDE_DN;
+		event->fxp = param;
+		break;
+	case 4:					/* Volume Slide Down */
+		event->fxt = FX_VOLSLIDE_DN;
+		event->fxp = param;
+		break;
+
+	case 10:				/* Fine Porta Up */
+		event->fxt = FX_F_PORTA_UP;
+		event->fxp = param;
+		break;
+	case 11:				/* Portamento Up */
+		event->fxt = FX_PORTA_UP;
+		event->fxp = param;
+		break;
+	case 12:				/* Fine Porta Down */
+		event->fxt = FX_F_PORTA_DN;
+		event->fxp = param;
+		break;
+	case 13:				/* Portamento Down */
+		event->fxt = FX_PORTA_DN;
+		event->fxp = param;
+		break;
+	case 14:				/* Tone Portamento */
+		event->fxt = FX_TONEPORTA;
+		event->fxp = param;
+		break;
+	case 15:				/* Glissando control */
+		event->fxt = FX_EXTENDED;
+		event->fxp = (EX_GLISS << 4) | (param & 0x0f);
+		break;
+	case 16:				/* Tone Portamento + Volslide Up */
+		event->fxt = FX_TONEPORTA;
+		event->fxp = 0;
+		event->f2t = FX_VOLSLIDE_UP;
+		event->f2p = param;
+		break;
+	case 17:				/* Tone Portamento + Volslide Down */
+		event->fxt = FX_TONEPORTA;
+		event->fxp = 0;
+		event->f2t = FX_VOLSLIDE_DN;
+		event->f2p = param;
+		break;
+
+	case 20:				/* Vibrato */
+		event->fxt = FX_VIBRATO;
+		event->fxp = param;
+		break;
+	case 21:				/* Vibrato waveform */
+		event->fxt = FX_EXTENDED;
+		event->fxp = (EX_VIBRATO_WF << 4) | (param & 0x0f);
+		break;
+	case 22:				/* Vibrato + Volume Slide Up */
+		event->fxt = FX_VIBRATO;
+		event->fxp = 0;
+		event->f2t = FX_VOLSLIDE_UP;
+		event->f2p = param;
+		break;
+	case 23:				/* Vibrato + Volume Slide Down */
+		event->fxt = FX_VIBRATO;
+		event->fxp = 0;
+		event->f2t = FX_VOLSLIDE_DN;
+		event->f2p = param;
+		break;
+
+	case 30:				/* Tremolo */
+		event->fxt = FX_TREMOLO;
+		event->fxp = param;
+		break;
+	case 31:				/* Tremolo waveform */
+		event->fxt = FX_EXTENDED;
+		event->fxp = (EX_TREMOLO_WF << 4) | (param & 0x0f);
+		break;
+
+	case 40:				/* Sample Offset */
```

Comment:
It's not used in any "official" PSMs but this effect is just as cursed as in the newer PSM format: The parameter is three bytes long, and the middle byte is the one that is equivalent to MOD/S3M/etc. offset. You can verify this by taking e.g. `song0.psm` from Epic Pinball, replace the byte at offset `0xED` (original value: `0x3C`, set speed) with `0x28` and you will notice that the menu music will skip a beat and then hang at the end of the pattern in-game, just like it will when throwing the modified file at OpenMPT.

Reply:
oh no :(